docs/recipes/reducers/03-RefactoringReducersExample.md

```diff
@@ -0,0 +1,420 @@
+
+```js
+function updateObject(oldObject, newValues) {
+    // Copy fields from oldObject to a new empty object.
+    // Then copy fields from newValues to the new object, overwriting any existing fields.
+    // This gives us a new updated copy, without modifying the original
+    return Object.assign({}, oldObject, newValues);
+}
+
+function updateItemInArray(array, itemId, updateItemCallback) {
+    const updatedItems = array.map(item => {
+        if(item.id !== itemId) {
+            // Since we only want to update one item, preserve all others as they are now
+            return item;
+        }
+        
+        // Use the provided callback to create an updated item
+        const updatedItem = updateItemCallback(item);
+        return updatedItem;
+    });
+    
+    return updatedItems;
+}
+
+
+function setVisibilityFilter(visibilityState, action) {
+    // Technically, we don't even care about the previous state
+    return action.filter;
+}
+
+function visibilityReducer(visibilityState = 'SHOW_ALL', action) {
+    switch(action.type) {
+        case 'SET_VISIBILITY_FILTER' : return setVisibilityFilter(visibilityState, action);
+        default : return visibilityState;
+    }
+};
+
+
+function addTodo(todosState, action) {
+    const newTodos = todosState.concat({
+        id: action.id,
+        text: action.text,
+        completed: false
+    });
+    
+    return newTodos;
+}
+
+function toggleTodo(todosState, action) {
+    const newTodos = updateItemInArray(todosState, action.id, todo => {
+        return updateObject(todo, {completed : !todo.completed});
+    });
+    
+    return newTodos;
+}
+
+function editTodo(todosState, action) {
+    const newTodos = updateItemInArray(todosState, action.id, todo => {
+        return updateObject(todo, {text : action.text});
+    });
+    
+    return newTodos;
+}
+
+function todosReducer(todosState = [], action) {
+    switch(action.type) {
+        case 'ADD_TODO' : return addTodo(todosState, action);
+        case 'TOGGLE_TODO' : return toggleTodo(todosState, action);
+        case 'EDIT_TODO' : return editTodo(todosState, action);
+        default : return todosState;
+    }
+}
+
+function appReducer(state = initialState, action) {
+    return {
+        todos : todosReducer(state.todos, action),
+        visibilityFilter : visibilityReducer(state.visibilityFilter, action)
+    };
+}
+
+```
+
+Notice that because the two "slice of state" reducers are now getting only their own part of the whole state as arguments, they no longer need to return complex nested state objects, and are now simpler as a result.
+
+Now, we can use Redux's built-in `combineReducers` utility to handle the "slice-of-state" logic.  Also, since the "per-action" reducer functions are following a consistent pattern, and many people don't like using switch statements, we can define a helper function that lets us organize them with a lookup table:
```

Comment:
I'm never sure whether it's best to repeat all the code every time, or to just show the changed code.  I struggle with that on my blog all the time. I tend to show just the changed code in order to cut down the length, but I've never had any feedback about which is easier to understand.

Perhaps you could show the initial code, then just the changes, then a final review of the ending code?